File: android/app/src/main/kotlin/games/yandex/wrap/catalog/CatalogDtoMappers.kt
package games.yandex.wrap.catalog

import games.yandex.wrap.catalog.models.AppDetail
import games.yandex.wrap.catalog.models.FeedBlock
import games.yandex.wrap.catalog.models.FeedPage
import games.yandex.wrap.catalog.models.FeedWithBlocks
import games.yandex.wrap.catalog.models.Game
import games.yandex.wrap.catalog.models.GameCategory
import games.yandex.wrap.catalog.models.UserProfile
import games.yandex.wrap.network.dtos.FeedBlockDto
import games.yandex.wrap.network.dtos.FeedResponseDto
import games.yandex.wrap.network.dtos.GameDto
import games.yandex.wrap.network.dtos.GetGameResponseDto
import games.yandex.wrap.network.dtos.ProfileResponseDto
import games.yandex.wrap.network.dtos.SimilarGamesResponseDto
import games.yandex.wrap.network.dtos.TagsResponseDto

internal fun FeedResponseDto.toFeedWithBlocks(): FeedWithBlocks {
    val blocks = feed.map { it.toDomain() }
    val topLevelGames = items.map(GameDto::toDomain)
    val nextPageId = pageInfo?.nextPageId ?: pageID
    return FeedWithBlocks(
        blocks = blocks,
        flatGames = blocks.flatMap { it.items }.ifEmpty { topLevelGames },
        recentGames = recentGames.map { it.toDomain() },
        genres = emptyList(),
        nextPageId = nextPageId,
        hasNext = pageInfo?.hasNextPage ?: (nextPageId != null || (totalPages ?: 0) > 1),
    )
}

internal fun FeedResponseDto.toFeedPage(): FeedPage {
    val nextPageId = pageInfo?.nextPageId ?: pageID
    return FeedPage(
        games = feed.flatMap { it.items.map(GameDto::toDomain) }.ifEmpty { items.map(GameDto::toDomain) },
        nextPageId = nextPageId,
        hasNext = pageInfo?.hasNextPage ?: (nextPageId != null || (totalPages ?: 0) > 1),
    )
}

internal fun SimilarGamesResponseDto.toGames(): List<Game> =
    games?.map { it.toDomain() } ?: feed.flatMap { it.items.map(GameDto::toDomain) }

internal fun GetGameResponseDto.toAppDetail(): AppDetail? {
    val game = game ?: return null
    return AppDetail(
        description = game.description,
        screenshots = game.media?.screenshots.orEmpty()
            .values
            .flatten()
            .mapNotNull { it.prefixUrl ?: it.url },
        datePublished = game.datePublished,
        genres = game.categoriesNames,
        languages = game.inLanguage,
        author = game.developer?.name,
    )
}

internal fun TagsResponseDto.toCategories(): List<GameCategory> = tags.mapNotNull { tag ->
    val slug = tag.slug.takeIf { it.isNotBlank() } ?: return@mapNotNull null
    val title = tag.title.takeIf { it.isNotBlank() } ?: return@mapNotNull null
    GameCategory(
        name = slug,
        title = title,
        gamesCount = tag.info?.gamesCount ?: 0,
    )
}

internal fun ProfileResponseDto.toUserProfile(): UserProfile? {
    if (uid.isBlank()) return null
    return UserProfile(
        isAuthorized = true,
        displayName = displayName.ifBlank { login },
        login = login,
        avatarUrl = avatarUrl(),
        hasYaPlus = yaplusEnabled,
    )
}

private fun ProfileResponseDto.avatarUrl(): String {
    if (avatarsOrigin.isBlank() || avatarId.isBlank()) return ""
    return "${avatarsOrigin.trimEnd('/')}/get-yapic/$avatarId/islands-200"
}

internal fun FeedBlockDto.toDomain(): FeedBlock = FeedBlock(
    type = type,
    size = size,
    title = title,
    items = items.map(GameDto::toDomain),
)

internal fun GameDto.toDomain(): Game {
    val coverPrefix = media?.cover?.prefixUrl
    val iconPrefix = media?.icon?.prefixUrl
    return Game(
        appId = appID,
        title = title,
        rating = rating,
        ratingCount = ratingCount,
        coverUrl = coverPrefix.orEmpty(),
        iconUrl = iconPrefix ?: coverPrefix.orEmpty(),
        categories = categoriesNames,
        developer = developer?.name.orEmpty(),
        mainColor = media?.cover?.mainColor,
        iconMainColor = media?.icon?.mainColor,
        videoUrl = media?.videos?.firstOrNull()?.mp4StreamUrl,
        coverPrefixUrl = coverPrefix,
        ageRating = features?.ageRating,
    )
}

File: ios/UGames/Catalog/CatalogDtoMappers.swift
import Foundation

extension FeedResponseDTO {
    var feedWithBlocks: FeedWithBlocks {
        let mappedBlocks = feed.orEmpty.map { $0.domain }
        let topLevelGames = items.orEmpty.map { $0.domain }
        let nextPageId = pageInfo?.nextPageId ?? pageID
        return FeedWithBlocks(
            blocks: mappedBlocks,
            flatGames: mappedBlocks.flatMap { $0.items }.isEmpty ? topLevelGames : mappedBlocks.flatMap { $0.items },
            recentGames: recentGames.orEmpty.map { $0.domain },
            genres: [],
            nextPageId: nextPageId,
            hasNext: pageInfo?.hasNextPage ?? (nextPageId != nil || (totalPages ?? 0) > 1)
        )
    }

    var feedPage: FeedPage {
        let nextPageId = pageInfo?.nextPageId ?? pageID
        let feedGames = feed.orEmpty.flatMap { $0.items.orEmpty.map { $0.domain } }
        return FeedPage(
            games: feedGames.isEmpty ? items.orEmpty.map { $0.domain } : feedGames,
            nextPageId: nextPageId,
            hasNext: pageInfo?.hasNextPage ?? (nextPageId != nil || (totalPages ?? 0) > 1)
        )
    }
}

extension SimilarGamesResponseDTO {
    var gamesDomain: [Game] {
        if let games {
            return games.map { $0.domain }
        }
        return feed.orEmpty.flatMap { $0.items.orEmpty.map { $0.domain } }
    }
}

extension TagsResponseDTO {
    var categories: [GameCategory] {
        tags.orEmpty.compactMap { tag in
            guard !tag.slug.isEmpty, !tag.title.isEmpty else { return nil }
            return GameCategory(name: tag.slug, title: tag.title, gamesCount: tag.info?.gamesCount ?? 0)
        }
    }
}

extension GetGameResponseDTO {
    var appDetail: AppDetail? {
        guard let game else { return nil }
        return AppDetail(
            description: game.description,
            screenshots: game.media?.screenshots.orEmpty.values.flatMap { $0 }.compactMap { $0.prefixUrl ?? $0.url } ?? [],
            datePublished: game.datePublished,
            genres: game.categoriesNames.orEmpty,
            languages: game.inLanguage.orEmpty,
            author: game.developer?.name
        )
    }
}

extension ProfileResponseDTO {
    var userProfile: UserProfile? {
        guard let uid, !uid.isEmpty else { return nil }
        return UserProfile(
            isAuthorized: true,
            displayName: (displayName ?? "").isEmpty ? (login ?? "") : (displayName ?? ""),
            login: login ?? "",
            avatarUrl: avatarUrl,
            hasYaPlus: yaplusEnabled ?? false
        )
    }

    private var avatarUrl: String {
        guard let avatarsOrigin, !avatarsOrigin.isEmpty, let avatarId, !avatarId.isEmpty else { return "" }
        let origin = avatarsOrigin.hasSuffix("/") ? String(avatarsOrigin.dropLast()) : avatarsOrigin
        return "\(origin)/get-yapic/\(avatarId)/islands-200"
    }
}

extension FeedBlockDTO {
    var domain: FeedBlock {
        FeedBlock(type: type, size: size, title: title ?? "", items: items.orEmpty.map { $0.domain })
    }
}

extension GameDTO {
    var domain: Game {
        let coverPrefix = media?.cover?.prefixUrl
        let iconPrefix = media?.icon?.prefixUrl
        return Game(
            appId: appID,
            title: title,
            rating: rating ?? 0,
            ratingCount: ratingCount ?? 0,
            coverUrl: coverPrefix ?? "",
            iconUrl: iconPrefix ?? coverPrefix ?? "",
            categories: categoriesNames.orEmpty,
            developer: developer?.name ?? "",
            mainColor: media?.cover?.mainColor,
            iconMainColor: media?.icon?.mainColor,
            videoUrl: media?.videos.orEmpty.first?.mp4StreamUrl,
            coverPrefixUrl: coverPrefix,
            ageRating: features?.ageRating
        )
    }
}
